------------------------------------------------------------
revno: 406 [merge]
fixes bug(s): https://launchpad.net/bugs/802637
committer: Scott Moser <smoser@ubuntu.com>
branch nick: trunk
timestamp: Thu 2011-06-30 13:57:42 -0400
message:
  [Marc Cluet] sanitize hosts file for system hostname (LP: #802637)
------------------------------------------------------------
Use --include-merges or -n0 to see merged revisions.
=== modified file 'ChangeLog'
--- old/ChangeLog	2011-06-28 10:41:05 +0000
+++ new/ChangeLog	2011-06-30 17:57:42 +0000
@@ -24,6 +24,7 @@
  - read /etc/ssh/sshd_config for AuthorizedKeysFile rather than
    assuming ~/.ssh/authorized_keys (LP: #731849)
  - fix cloud-init in ubuntu lxc containers (LP: #800824)
+ - sanitize hosts file for system's hostname to 127.0.1.1 (LP: #802637)
 0.6.1:
  - fix bug in fixing permission on /var/log/cloud-init.log (LP: #704509)
  - improve comment strings in rsyslog file tools/21-cloudinit.conf

=== modified file 'cloudinit/CloudConfig/cc_update_etc_hosts.py'
--- old/cloudinit/CloudConfig/cc_update_etc_hosts.py	2011-02-17 21:22:53 +0000
+++ new/cloudinit/CloudConfig/cc_update_etc_hosts.py	2011-06-30 17:35:12 +0000
@@ -17,12 +17,44 @@
 #    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 import cloudinit.util as util
 from cloudinit.CloudConfig import per_always
+import platform
+import StringIO
 
 frequency = per_always
 
 def handle(name,cfg,cloud,log,args):
     if not util.get_cfg_option_bool(cfg,"manage_etc_hosts",False):
-        log.debug("manage_etc_hosts is not set. not modifying /etc/hosts")
+        log.debug("manage_etc_hosts is not set, checking sanity of /etc/hosts")
+        with open('/etc/hosts', 'r') as etchosts:
+            current_hostname = platform.node()
+            hosts_line = "# Added by cloud-init\n127.0.1.1\t%s.localdomain %s\n" % (current_hostname, current_hostname)
+            need_write = False
+            need_change = True
+            new_etchosts = StringIO.StringIO()
+            for line in etchosts:
+                split_line = [s.strip() for s in line.split()]
+                # skip over malformed /etc/hosts entries
+                if len(split_line) < 2:
+                    continue
+                ip, hostnames = split_line[0], split_line[1:]
+                if ip == "127.0.1.1":
+                    for hostname in hostnames:
+                        if hostname == current_hostname:
+                            need_change = False
+                    if need_change == True:
+                        line = hosts_line
+                        need_change = False
+                        need_write = True
+                new_etchosts.write(line)
+            etchosts.close()
+            if need_change == True:
+                new_etchosts.write(hosts_line)
+                need_write = True
+            if need_write == True:
+                new_etcfile = open ('/etc/hosts','wb')
+                new_etcfile.write(new_etchosts.getvalue())
+                new_etcfile.close()
+            new_etchosts.close()
         return
 
     try:

